draft-ietf-tls-tls13.md

```diff
@@ -2349,12 +2350,16 @@ RSASSA-PSS algorithms
   digest used in the mask generation function and the digest being signed are
   both the corresponding hash algorithm as defined in {{SHS}}. When used in
   signed TLS handshake messages, the length of the salt MUST be equal to the
-  length of the digest output. An RSA public key (OID rsaEncryption) MUST be
-  supported, an RSASSA-PSS public key (OID id-RSASSA-PSS) MAY be supported.
-  When used as signature algorithm in certificates, the same parameters as used
-  for TLS handshake messages MUST be supported but other parameters MAY be used.
+  length of the digest output.
+  When used as signature algorithm in certificates, the length of the salt
+  SHOULD be equal to the length of the digest output, but other salt lengths MAY
```

Comment:
> but other salt lengths MAY be used

I find that text a step backwards in terms of interoperability. The previous text of requiring that the salt length MUST be equal to the length of the digest output was specific and left no window of misunderstandings.

Reply:
The intention is that use of a standard length is guaranteed, but if a CA somehow issues certs with odd salt lengths, then TLS would not outright forbid it (and leave it up to the PKI to perform policy decisions). See also https://github.com/tlswg/tls13-spec/pull/1109#discussion_r153673401

It was added based on comments from Martin Rex (I am not strongly in favor in keeping this freedom though).

Reply:
It is a slippery slope. TLS currently specifies exactly the combo of parameters to use for RSA-PSS signing. The negotiation includes the hash and the salt size. What the next text makes, is make the TLS negotiation include only the hash algorithm, and leaves the application to deal with the salt size requirements outside TLS. I find the previous text simpler to implement and in alignment with having MUST in certificates with no parameters.

Reply:
how about "but longer salt lengths MAY be used"? AFAIR OpenSSL defaults to maximum allowed size (by public key size) of salt for rsa-pss signatures.

I think it would be silly to reject those certificates only because the salt is too large

Reply:
Note that PKCS#1 v2.2 (https://tools.ietf.org/html/rfc8017#page-42) seems to suggest that zero-length salts are also okay:
>    4.  Typical salt lengths in octets are hLen (the length of the output
>       of the hash function Hash) and 0.  In both cases, the security of
>       RSASSA-PSS can be closely related to the hardness of inverting
>       RSAVP1.

Maybe requiring a single salt length is the best for interop purposes. Btw, I think I'm going to close this in favor of #1109, let's continue the discussion there.

Reply:
I have double-checked now - OpenSSL does default to maximum salt length possible for a given key size when making signatures - so requiring a salt length of the same size as the hash length would make all OpenSSL certificates made with default PSS settings invalid.

while 0 length salts are ok in RFC 8017, using different hash for MGF1 and signature hash is also ok, so I don't think that's the sole data point we will want to use